cpp/include/graph.hpp

```diff
@@ -214,5 +214,182 @@ class GraphCSC: public GraphCompressedSparseBase<VT,ET,WT> {
   {}
 };
 
+/**
+ * @brief       Base class for graphs constructed in the C++ API
+ *
+ * This class will own edge data, until the data is moved.
+ *
+ * @tparam VT   Type of vertex id
+ * @tparam ET   Type of edge id
+ * @tparam WT   Type of weight
+ */
+template <typename VT, typename ET, typename WT>
+class ConstructedGraphBase {
+public:
+  std::unique_ptr<WT>  edge_data{};     ///< edge weight
```

Comment:
Why unique_ptr instead of rmm::device_buffer?

First, if you are storing more than one weights, this should be std:::unique_ptr<WT[]>. And if this memory is allocated on GPU through cudaMalloc, custom deleter should appear in the declaration. AFAIK, std::make_unique (which is exception safe) does not work with a custom deleter. So, if you need to allocate using cudaMalloc and create a unique_ptr object and this leaves a hole in exception safety. And this escapes our rmm based memory management policy.

So, is there really a reason to go for a unique_ptr considering all these complexities?

Reply:
@kaatish has created a separate PR #799 that I think supersedes this idea.  He took this idea and has expanded it based upon a bunch of discussion on using rmm:device_buffer.  I should probably delete this PR, but was leaving it in place so I could look back at it while comparing to his.